src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs

```diff
@@ -189,10 +203,16 @@ public MemoryPoolIterator ConsumingStart()
                 {
                     if (!consumed.IsDefault)
                     {
+                        var lengthConsumed = new MemoryPoolIterator(_head).GetLength(consumed);
+
                         returnStart = _head;
                         returnEnd = consumed.Block;
                         _head = consumed.Block;
                         _head.Start = consumed.Index;
+
+                        // Must call Subtract() after bytes have been freed, to avoid producer starting too early and growing
+                        // buffer beyond max length.
```

Comment:
This comment implies that the bytes should already be returned to the pool which they're not. Maybe "after _head as been advanced" would be clearer? Even that I don't think is too important as long as `Subtract` is called with the `_sync` lock.